File: union_find/ruby/lib/UF/quick_find.rb
class UF::QuickFind
  attr_reader :id

  def initialize(n)
    @n = n.to_i
    @id = (0..@n-1).to_a
  end

  def union(p, q)
    return if self.connected?(p, q)
    pid = self.id[p]
    qid = self.id[q]

    @id = self.id.map do |idEl|
      idEl == pid ? qid : idEl
    end
  end

  def connected?(p, q)
    self.id[p] == self.id[q]
  end
end

File: union_find/javascript/QuickFind.js
class QuickFind {

  constructor(N) {
    var i
    this.id = []
    for(i = 0; i < N; i++) {
      this.id.push(i)
    }
  }

  connected(p, q) {
    return this.id[p] == this.id[q];
  }

  union(p, q) {
    var i;
    var pid = this.id[p];
    var qid = this.id[q];
    for(i = 0; i < this.id.length; i++) {
      if(pid == this.id[i]) this.id[i] = qid;
    }

    console.log(this.id)
  }
}
